mm2src/coins/sia/src/encoding.rs

```diff
@@ -0,0 +1,345 @@
+        let decoded_vec = hex::decode(hex_str).map_err(serde::de::Error::custom)?;
+
+        if decoded_vec.len() != 64 {
+            return Err(serde::de::Error::custom("Invalid length: expected 64 byte hex string"));
+        }
+
+        let array: [u8; 64] = decoded_vec
+            .try_into()
+            .map_err(|_| serde::de::Error::custom("Failed to convert Vec<u8> to [u8; 64]"))?;
+
+        Ok(HexArray64(array))
+    }
+}
+
+impl Serialize for HexArray64 {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let hex_str = hex::encode(self.0);
+        serializer.serialize_str(&hex_str)
+    }
+}
+
+// https://github.com/SiaFoundation/core/blob/092850cc52d3d981b19c66cd327b5d945b3c18d3/types/encoding.go#L16
+// TODO go implementation limits this to 1024 bytes, should we?
+#[derive(Default)]
+pub struct Encoder {
+    pub buffer: Vec<u8>,
+}
+
+pub trait Encodable {
+    fn encode(&self, encoder: &mut Encoder);
+}
+
+// This wrapper allows us to use Signature internally but still serde as "sig:" prefixed string
+#[derive(Debug)]
+pub struct PrefixedSignature(pub Signature);
+
+impl<'de> Deserialize<'de> for PrefixedSignature {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct PrefixedSignatureVisitor;
+
+        impl<'de> serde::de::Visitor<'de> for PrefixedSignatureVisitor {
+            type Value = PrefixedSignature;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a string prefixed with 'sig:' and followed by a 128-character hex string")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                if let Some(hex_str) = value.strip_prefix("sig:") {
+                    Signature::from_str(hex_str)
+                        .map(PrefixedSignature)
+                        .map_err(|_| E::invalid_value(serde::de::Unexpected::Str(value), &self))
+                } else {
+                    Err(E::invalid_value(serde::de::Unexpected::Str(value), &self))
+                }
+            }
+        }
+
+        deserializer.deserialize_str(PrefixedSignatureVisitor)
+    }
+}
+
+impl Serialize for PrefixedSignature {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&format!("{}", self))
+    }
+}
+
+impl fmt::Display for PrefixedSignature {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result { write!(f, "sig:{:x}", self.0) }
+}
+
+impl From<PrefixedSignature> for Signature {
+    fn from(sia_hash: PrefixedSignature) -> Self { sia_hash.0 }
+}
+
+impl From<Signature> for PrefixedSignature {
+    fn from(signature: Signature) -> Self { PrefixedSignature(signature) }
+}
```

Comment:
@Alrighttt please make sure the fix here https://github.com/KomodoPlatform/komodo-defi-framework/pull/2167/commits/d6e734ce7af3759cd5f4a084355c426c33882dfe is fine

Reply:
There is a test failing seemingly due to this commit. Checking now
https://github.com/KomodoPlatform/komodo-defi-framework/actions/runs/10571920147/job/29288799784

I almost prefer to leave these alone at this point. I dislike large macros like this because they hard to read at a glance.

Reply:
Seems this commits change the serde of UnlockCondition 

```
                        "policy": {
                            "publicKeys": [
                                "ed25519:cecc1507dc1ddd7295951c290888f095adb9044d1b73d696e6df065d683bd4fc"
                            ],
                            "timelock": 0,
                            "signaturesRequired": 1
                        }
```

vs

```
                        "policy": {
                            "timelock": 0,
                            "publicKeys": [
                                "ed25519:cecc1507dc1ddd7295951c290888f095adb9044d1b73d696e6df065d683bd4fc"
                            ],
                            "signaturesRequired": 1
                        }
```

Reply:
> I dislike large macros like this because they hard to read at a glance.

I use macro expansion as below to read the implementations, but it's fine by me. I can revert the last commit
<img width="532" alt="Screenshot 2024-08-27 at 4 04 24 PM" src="https://github.com/user-attachments/assets/f88df2f4-a8a7-4937-8f48-cc1866e809a0">

Reply:
Reverted